flaky test spotted in CI

Sometimes it will fail for some reason, need to check.

```txt
	/nix/store/kyai650rrywrzflnwda9mbk2c2j9w68r-go-1.26.4/share/go/src/testing/testing.go:2036
2026-07-26T13:44:15.413Z	DEBUG	hotkeys	hotkeys/manager_darwin.go:199	Setting global hotkey manager to nil
2026-07-26T13:44:15.413Z	DEBUG	hotkeys	hotkeys/manager_darwin.go:146	Unregistering all hotkeys
2026-07-26T13:44:15.413Z	DEBUG	hotkeys	hotkeys/manager_darwin.go:157	Unregistered all hotkeys
--- FAIL: TestHotkeyAdapterIntegration (10.10s)
    --- FAIL: TestHotkeyAdapterIntegration/Register_and_Unregister (5.05s)
    --- PASS: TestHotkeyAdapterIntegration/Register_Invalid_Hotkey (5.05s)
=== RUN   TestNewAdapter
--- PASS: TestNewAdapter (0.00s)
=== RUN   TestAdapter_Register
--- PASS: TestAdapter_Register (0.00s)
```

--- a/docs/testing/TESTING_PATTERNS.md
+++ b/docs/testing/TESTING_PATTERNS.md
@@ -102,6 +102,32 @@ func TestDarwinAccessibility(t *testing.T) {
 }
 ```
 
+#### macOS tests that need the main run loop
+
+Native work such as building the keyboard layout maps or creating a `CGEventTap`
+is dispatched to the main queue, which is only drained while the main run loop
+runs. The daemon starts that loop in `cmd/neru`; a `go test` binary never does.
+A test that reaches such code therefore only passes when the Go scheduler
+happens to run it on the main OS thread — otherwise `dispatch_async` work
+silently times out (an empty keymap makes every key name fail to parse) and
+`dispatch_sync` deadlocks.
+
+Pump the run loop from `TestMain` so the behaviour is deterministic:
+
+```go
+// Pin the main thread during package init so TestMain still runs on it.
+func init() {
+    runtime.LockOSThread()
+}
+
+func TestMain(m *testing.M) {
+    os.Exit(darwin.RunMainLoopForTesting(m.Run))
+}
+```
+
+`*_integration_darwin_test.go` files are exempt from the "One Rule", so they may
+import `internal/core/infra/platform/darwin` directly.
+
 ### Test Command Usage
 
 - `just test`: Runs all unit tests (platform-agnostic).

--- a/internal/core/infra/eventtap/adapter_integration_darwin_test.go
+++ b/internal/core/infra/eventtap/adapter_integration_darwin_test.go
@@ -4,13 +4,27 @@ package eventtap_test
 
 import (
 	"context"
+	"os"
+	"runtime"
 	"testing"
 
 	"github.com/y3owk1n/neru/internal/core/infra/eventtap"
 	"github.com/y3owk1n/neru/internal/core/infra/logger"
-	_ "github.com/y3owk1n/neru/internal/core/infra/platform/darwin" // Link CGO implementations
+	"github.com/y3owk1n/neru/internal/core/infra/platform/darwin" // Link CGO implementations
 	"github.com/y3owk1n/neru/internal/core/ports"
 )
+
+// Pin the main thread during package init so TestMain still runs on it.
+func init() {
+	runtime.LockOSThread()
+}
+
+// TestMain services the macOS main run loop while the tests run. Resolving
+// hotkey strings to key codes happens on the main queue, which nothing drains
+// in a plain `go test` binary.
+func TestMain(m *testing.M) {
+	os.Exit(darwin.RunMainLoopForTesting(m.Run))
+}
 
 // TestEventTapAdapterImplementsPort verifies the adapter implements the port interface.
 func TestEventTapAdapterImplementsPort(_ *testing.T) {

--- a/internal/core/infra/hotkeys/adapter_integration_darwin_test.go
+++ b/internal/core/infra/hotkeys/adapter_integration_darwin_test.go
@@ -4,13 +4,28 @@ package hotkeys_test
 
 import (
 	"context"
+	"os"
+	"runtime"
 	"testing"
 
 	"github.com/y3owk1n/neru/internal/core/infra/hotkeys"
 	"github.com/y3owk1n/neru/internal/core/infra/logger"
-	_ "github.com/y3owk1n/neru/internal/core/infra/platform/darwin" // Link CGO implementations
+	"github.com/y3owk1n/neru/internal/core/infra/platform/darwin" // Link CGO implementations
 	"github.com/y3owk1n/neru/internal/core/ports"
 )
+
+// Pin the main thread during package init so TestMain still runs on it.
+func init() {
+	runtime.LockOSThread()
+}
+
+// TestMain services the macOS main run loop while the tests run. Registering a
+// hotkey needs the keyboard layout maps and a CGEventTap, both of which are
+// built on the main queue; nothing drains that queue in a plain `go test`
+// binary.
+func TestMain(m *testing.M) {
+	os.Exit(darwin.RunMainLoopForTesting(m.Run))
+}
 
 // TestHotkeyAdapterImplementsPort verifies the adapter implements the port interface.
 func TestHotkeyAdapterImplementsPort(_ *testing.T) {

--- a/internal/core/infra/platform/darwin/keymap_darwin.m
+++ b/internal/core/infra/platform/darwin/keymap_darwin.m
@@ -895,6 +895,15 @@ static void waitForLayoutMapsInitialized(NSTimeInterval timeoutSeconds) {
 		}
 	}
 	[gLayoutInitCondition unlock];
+
+	// The build is dispatched to the main queue, so a timeout means the main
+	// run loop never ran it. Callers then see an empty keymap and every key
+	// name fails to resolve, which is worth saying out loud.
+	if (!atomic_load_explicit(&gLayoutMapsInitialized, memory_order_acquire)) {
+		NSLog(
+		    @"Neru: timed out after %.1fs waiting for keyboard layout maps (main run loop not running?)",
+		    timeoutSeconds);
+	}
 }
 
 /// Register notification observer (called once from initializeKeyMaps)

--- a/internal/core/infra/platform/darwin/runloop_darwin.go
+++ b/internal/core/infra/platform/darwin/runloop_darwin.go
@@ -1,0 +1,72 @@
+//go:build darwin
+
+package darwin
+
+/*
+#include <CoreFoundation/CoreFoundation.h>
+
+// Runs the main run loop for at most `seconds` and returns the CFRunLoopRunResult.
+static int neruRunMainLoopSlice(double seconds) {
+	return (int)CFRunLoopRunInMode(kCFRunLoopDefaultMode, seconds, false);
+}
+
+static void neruStopMainLoop(void) {
+	CFRunLoopStop(CFRunLoopGetMain());
+}
+*/
+import "C"
+
+import "time"
+
+// mainLoopSlice bounds how long a single run loop pump blocks before the
+// caller re-checks whether the tests finished.
+const mainLoopSlice = 100 * time.Millisecond
+
+// cfRunLoopRunFinished mirrors kCFRunLoopRunFinished: the mode had nothing left
+// to service, so CFRunLoopRunInMode returned without waiting out its timeout.
+const cfRunLoopRunFinished = 1
+
+// RunMainLoopForTesting runs testMain on a background goroutine while the
+// calling goroutine services the CoreFoundation main run loop, and returns the
+// exit code testMain produced.
+//
+// Native work Neru depends on — building the keyboard layout maps, creating a
+// CGEventTap — is dispatched to the main queue, which is only drained while the
+// main run loop runs. The daemon starts that loop (see cmd/neru), but a `go
+// test` binary never does, so without this helper those dispatches are never
+// serviced: dispatch_async work silently times out (leaving, for example, an
+// empty keymap so every key name fails to parse) and dispatch_sync deadlocks.
+// Whether a test hit that depended on which OS thread the Go scheduler happened
+// to run it on, which is what made the macOS integration tests flaky.
+//
+// Call it from TestMain, and lock the main thread from an init function in the
+// same package so TestMain is guaranteed to run on it:
+//
+//	func init() { runtime.LockOSThread() }
+//
+//	func TestMain(m *testing.M) { os.Exit(darwin.RunMainLoopForTesting(m.Run)) }
+func RunMainLoopForTesting(testMain func() int) int {
+	result := make(chan int, 1)
+
+	go func() {
+		result <- testMain()
+
+		// Cut the current pump short so the exit code is picked up
+		// immediately instead of after the remainder of the slice.
+		C.neruStopMainLoop()
+	}()
+
+	for {
+		select {
+		case code := <-result:
+			return code
+		default:
+		}
+
+		// A run loop with no sources left returns immediately; sleep out
+		// the slice instead of spinning on it.
+		if C.neruRunMainLoopSlice(C.double(mainLoopSlice.Seconds())) == cfRunLoopRunFinished {
+			time.Sleep(mainLoopSlice)
+		}
+	}
+}

